core-processor/src/processor.rs

```diff
@@ -42,33 +49,197 @@ enum SuccessfulDispatchResultKind {
         dispatch,
         origin,
         gas_allowance,
+        subsequent_execution,
     } = execution_context;
     let Actor {
         balance,
         destination_program,
         executable_data,
     } = actor;
 
-    match check_is_executable(executable_data, &dispatch) {
-        Err(exit_code) => process_non_executable(dispatch, destination_program, exit_code),
-        Ok(data) => process_executable::<A, E>(
-            origin,
-            gas_allowance,
-            data,
+    let data = match check_is_executable(executable_data, &dispatch) {
+        Err(exit_code) => {
+            return PrepareResult::Error(process_non_executable(
+                dispatch,
+                destination_program,
+                exit_code,
+            ))
+        }
+        Ok(data) => data,
+    };
+
+    let program = &data.program;
+    let program_id = program.id();
+    let mut gas_counter = GasCounter::new(dispatch.gas_limit());
+    if !program.code().exports().contains(&dispatch.kind()) {
+        return PrepareResult::WontExecute(process_success(
+            SuccessfulDispatchResultKind::Success,
+            DispatchResult::success(dispatch, program_id, gas_counter.into()),
+        ));
+    }
+
+    let mut gas_allowance_counter = GasAllowanceCounter::new(gas_allowance);
+    let memory_size = match executor::charge_gas_for_pages(
+        &block_config.allocations_config,
+        &mut gas_counter,
+        &mut gas_allowance_counter,
+        program.get_allocations(),
+        program.static_pages(),
+        dispatch.context().is_none() && matches!(dispatch.kind(), DispatchKind::Init),
+        subsequent_execution,
+    ) {
+        Ok(size) => size,
+        Err(reason) => {
+            log::debug!("failed to charge for memory pages: {:?}", reason);
+            return PrepareResult::Error(match reason {
+                ExecutionErrorReason::InitialMemoryBlockGasExceeded
+                | ExecutionErrorReason::GrowMemoryBlockGasExceeded
+                | ExecutionErrorReason::LoadMemoryBlockGasExceeded => {
+                    process_allowance_exceed(dispatch, program_id, gas_counter.burned())
+                }
+                _ => process_error(dispatch, program_id, gas_counter.burned(), reason),
+            });
+        }
+    };
+
+    PrepareResult::Ok {
+        context: PreparedMessageExecutionContext {
+            gas_counter,
+            gas_allowance_counter,
             dispatch,
+            origin,
             balance,
-            block_config.clone(),
-        ),
+            program: data.program,
+            memory_size,
+        },
+        pages_with_data: data.pages_with_data,
+    }
+}
+
+/// Process program & dispatch for it and return journal for updates.
+pub fn process<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environment<A>>(
+    block_config: &BlockConfig,
+    execution_context: PreparedMessageExecutionContext,
+    memory_pages: BTreeMap<PageNumber, PageBuf>,
```

Comment:
can't memory pages be thrown with prepared message context?

Reply:
No, they cannot since it is responsibility of `prepare`  caller to fetch memory pages from storage on success. So to pass memory data with `PreparedMessage...` there should be added getter/setters/etc and `execution_context` has to be mutable - looks ugly. As an alternative new type could be introduced composing from `memory_pages` + `PrepareMessage...` but again it creates unnecessary  complications. So I would go with the current solution.